Implement multiple serial requests

christophermanahan opened the pull request:
### Description

- Add the ability for the server to accept connections and respond to requests serially.

christophermanahan committed:
Add multiple serial requests acceptance test

christophermanahan committed:
Alter HTTPServer to close after a single request

christophermanahan committed:
Modify Main to accept multiple serial requests

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,12 @@
+            ServerSocket serverSocket = new ServerSocket(port);
+            new HTTPServer(new ServerSocketListener(serverSocket), new ErrorLogger()).run();
+            serverSocket.close();
+        }
```
Should `Main` be responsible for this logic?

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,12 @@
-        serverSocket = new ServerSocket(port);
-        new HTTPServer(new ServerSocketListener(serverSocket), new ErrorLogger()).run();
+        while (true) {
+            ServerSocket serverSocket = new ServerSocket(port);
```
What are the implications of creating a new `ServerSocket` on each iteration?

christophermanahan replied:
Creating a new ServerSocket would not allow multiple different clients to connect and send requests. I believe it would wipe out any connection requests sent to the ServerSocket while a request is currently being processed by the server. I will pull it outside of the loop.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,12 @@
-        new HTTPServer(new ServerSocketListener(serverSocket), new ErrorLogger()).run();
+        while (true) {
+            ServerSocket serverSocket = new ServerSocket(port);
+            new HTTPServer(new ServerSocketListener(serverSocket), new ErrorLogger()).run();
```
Is this really an `HTTPServer` anymore? I would expect a "server" to have its own loop.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,12 @@
+        while (true) {
+            ServerSocket serverSocket = new ServerSocket(port);
+            new HTTPServer(new ServerSocketListener(serverSocket), new ErrorLogger()).run();
+            serverSocket.close();
```
Do we need to close on each iteration?

christophermanahan replied:
No, I believe this is unnecessary after pulling the ServerSocket out of the loop as stated above.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/HTTPServer.java:
```diff
@@ -1,20 +1,20 @@
             connect();
-            serveUntilDisconnect();
+            serve();
+            close();
```
What would happen if an exception occurred before we got to this line?

christophermanahan replied:
The connection would not close. But close also throws an exception. Should I have another try catch after this? Or should I place close in the catch and not worry about its exception? This is making me think that close should potentially not throw an exception.

damonkelley commented on src/test/resources/gradle/cucumber/multipleRequests.feature:
```diff
@@ -0,0 +1,9 @@
+    When I send method "GET" for "/simple_get" to host at the specified port 2 times
+    Then I should receive responses with version "HTTP_1_1"
+    And Status codes 200
+    And Bodies ""
```
Is this test valuable to us?

christophermanahan replied:
No, this was leftover from the previous POST request attempt.

christophermanahan committed:
Modify server to use a single ServerSocket instance

christophermanahan committed:
Rewrite HTTPServer to use try with resources
Rename Listener to Acceptor, implement acceptor and connection to be autoclosable, and implement using try with resources.

christophermanahan committed:
Remove unused isOpen method from Connection

christophermanahan committed:
Remove unnecessary body from Cucumber scenarios

christophermanahan commented:
@damonkelley I believe I have addressed the current review comments.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Acceptor.java:
```diff
@@ -0,0 +1,9 @@
+package io.github.christophermanahan.carnitas;
+
+public interface Acceptor extends AutoCloseable {
```
I am not in love with this name, but not a blocker. Let's keep an eye out for something better as we move forward.

christophermanahan replied:
Me neither, this has proved quite the naming challenge.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/HTTPServer.java:
```diff
@@ -1,38 +1,28 @@
-            serve();
+    void run() {
+        while (acceptor.isAccepting()) {
+            try (Connection connection = acceptor.accept()) {
```
Nice!

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,22 @@
+          ServerSocket serverSocket = new ServerSocket(port);
+          Acceptor acceptor = new ConnectionAcceptor(serverSocket)
+        ) {
+            connectionAcceptor = acceptor;
```
Is this necessary?

damonkelley replied:
Oh I see what is going on here

damonkelley replied:
Let's make a card to come back and address this more elegantly later.

christophermanahan replied:
Sounds like a plan.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,12 +4,22 @@
+            connectionAcceptor = acceptor;
+            new HTTPServer(acceptor, new ErrorLogger()).run();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Do we need to rethrow a `RuntimeException` here?

christophermanahan replied:
No, IOException is now ignored.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -100,44 +97,53 @@ public void log(String message) {
 
-        public Connection listen() {
+        public TestConnection accept() {
+            receiveCount -= 1;
```
Do we need this?

christophermanahan replied:
No, switched over to an iterator.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -100,44 +97,53 @@ public void log(String message) {
 
-        public TestConnection(List<String> received, List<String> sent) {
-            this.received = received.iterator();
+        TestConnection(String received, List<String> sent) {
```
What is received?

christophermanahan replied:
Renamed to request.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/SocketConnectionTest.java:
```diff
@@ -64,23 +62,12 @@ void throwsExceptionIfCloseFails() {
         private boolean closed;
 
-        public TestSocket(String receive) {
+        TestSocket(String receive) {
```
What is receive?

christophermanahan replied:
Renamed to request.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -100,44 +97,53 @@ public void log(String message) {
-        public Connection listen() {
+        public TestConnection accept() {
+            connection.open();
+            receiveCount -= 1;
```
I have some ideas on how to make this a little more elegant. Maybe we can pair on that this Friday?

damonkelley replied:
But not a blocker for this PR.

christophermanahan replied:
Oh, well, we should still pair on it, but I switched it over to an iterator of connections after seeing the previous comment!

christophermanahan committed:
Remove runMain and configure gradle run

christophermanahan committed:
Refactor HTTPServerTest to use a Connection iterator

christophermanahan committed:
Rename SocketConnectionTest receive to request

christophermanahan committed:
Ignore IOException in Main

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -14,9 +14,7 @@ public static void main(String[] args) {
-        } catch (IOException e) {
-            throw new RuntimeException(e);
-        }
+        } catch (IOException ignored) {}
```
I think it is probably still a good idea to log that something happened. I just don't think we need to rethrow a runtime exception.

christophermanahan replied:
Logged.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -98,52 +95,50 @@ public void log(String message) {
-            this.received = received;
+        TestConnection(String request, List<String> sent) {
+            this.request = request;
             this.sent = sent;
```
What is `sent` used for?

christophermanahan replied:
Renamed to responses.

christophermanahan committed:
Rename HTTPServerTest sent to responses

christophermanahan committed:
Add error logging to Main

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -128,12 +128,12 @@ public TestConnection accept() {
+        TestConnection(String request, List<String> responses) {
             this.request = request;
-            this.sent = sent;
+            this.responses = responses;
```
Will a test connection ever send back multiple responses?
